src/pages/order-confirm/context/CouponProvider.tsx

```diff
@@ -0,0 +1,104 @@
+import {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { Cart } from "../../../api/cart";
+import { Coupon } from "../../../api/coupon";
+import { optimizeCouponSelection } from "../../../domain/coupon/utils/couponOptimizer";
+
+interface CouponContextInterface {
+  selectedCoupons: Coupon[];
+  handleCouponToggle: (coupon: Coupon) => void;
+  handleCouponSelection: (coupons: Coupon[]) => void;
+  initializeCoupons: (
+    coupons: Coupon[],
+    totalCartPrice: number,
+    shippingFee: number,
+    selectedCartItems: Cart[] | undefined
+  ) => boolean;
+}
+
+const CouponContext = createContext<CouponContextInterface | null>(null);
+
+export const CouponProvider = ({ children }: PropsWithChildren) => {
+  const [selectedCoupons, setSelectedCoupons] = useState<Coupon[]>([]);
+  const [isInitialized, setIsInitialized] = useState(false);
+
+  useEffect(() => {
+    return () => {
+      setIsInitialized(false);
+    };
+  }, []);
+
+  const handleCouponToggle = useCallback(
+    (coupon: Coupon) => {
+      setSelectedCoupons((prev) => {
+        if (prev.some((c) => c.id === coupon.id)) {
+          return prev.filter((c) => c.id !== coupon.id);
+        } else {
+          if (prev.length >= 2) {
+            return prev;
+          }
+          return [...prev, coupon];
+        }
+      });
```

Comment:
```suggestion
      setSelectedCouponIds(prev => {
    const next = new Set(prev);
    if (next.has(coupon.id)) {
      next.delete(coupon.id);
    } else if (next.size < 2) {
      next.add(coupon.id);
    }
    return next;
  });
```

요렇게도 가능할거 같은데 어떻게 생각하는지? 😁